tests/eas/st_hold.py

```diff
@@ -0,0 +1,409 @@
+        self.target = self.env.target
+        # check that the SCHEDTUNE_HOLD_ALL feature exists in the target
+        self.check_hold_available()
+        # check any per-class requirements
+        self.child_check_config()
+        self.rta_name = 'rt-test-{}'.format(self.task_name.replace('/',''))
+        # ask derived class to populate workload config etc.
+        self.populate_params()
+        # ensure required cgroups exist
+        self.setup_cgroups()
+        # run workload
+        with self.env.freeze_userspace():
+            self.run_workload()
+        # collect and parse the trace
+        self.ftrace_obj = FTrace(self.trace_file)
+
+    @classmethod
+    def child_check_config(self):
+        pass
+
+    @classmethod
+    def run_workload(self):
+        self.env.ftrace.start()
+        self.rtapp.run(out_dir=self.env.res_dir, cgroup=self.cgroup_name, as_root=True)
+        self.env.ftrace.stop()
+        self.env.ftrace.get_trace(self.trace_file)
+        self.env.platform_dump(self.env.res_dir)
+
+    @classmethod
+    def setup_cpuset_cgroup(self, cgroup, cpus, mems):
+        cpu_g = self.cpuset_controller.cgroup(cgroup)
+        cpu_g.set(cpus=cpus, mems=mems)
+
+    @classmethod
+    def setup_schedtune_cgroup(self, cgroup, boost, prefer_idle):
+        st_g = self.schedtune_controller.cgroup(cgroup)
+        st_g.set(boost=boost, prefer_idle=prefer_idle)
+
+    @classmethod
+    def setup_cgroups(self):
+        """
+        When we set up the cgroups in a unified hierarchy, we
+        need to ensure that the cpu controller is minimally
+        configured before we can run tasks inside a group.
+        Copy the root group's mems and cpus parameters into the
+        groups we create just in case.
+        """
+        self.cpuset_controller = self.target.cgroups.controller('cpuset')
+        root_g = self.cpuset_controller.cgroup('/')
+        params = root_g.get()
+        self.schedtune_controller = self.target.cgroups.controller('schedtune')
+        self.setup_schedtune_cgroup(self.boosted_cgroup_name, 100, 0)
+        self.setup_cpuset_cgroup(self.boosted_cgroup_name,
+                                params['cpus'], params['mems'])
+        self.setup_schedtune_cgroup(self.unboosted_cgroup_name, 0, 0)
+        self.setup_cpuset_cgroup(self.unboosted_cgroup_name,
+                                params['cpus'], params['mems'])
+
+    @classmethod
+    def get_task_times(self):
+        """
+        figure out timestamps where task is running
+        defined here as:
+          start_time = first time we switch to the correctly named task
+          end_time   = last time we switch away from the correctly named task
+        """
+        df = self.ftrace_obj.sched_switch.data_frame
+        self.task_start_time = df[df.next_comm == self.task_name].iloc[0].name
+        self.task_end_time = df[df.prev_comm == self.task_name].iloc[-1].name
+
+    @classmethod
+    def avg_cpu_freq(self):
```

Comment:
Is it possible to simplify this one a bit ? I found the following under `analysis/frequency_analysis.py`:

```python
    avg_freq = 0
    if len(_df) > 1:
        timespan = _df.index[-1] - _df.index[0]
        avg_freq = area_under_curve(_df['frequency'], method='rect') / timespan
 ```

`area_under_curve` comes from bart.

Reply:
Maybe, I'll take a look

Reply:
It doesn't seem to be able to figure out the area under the curve of an arbitrary slice of data, only a whole series - I could use it, but I think I'd have to build a new series using the frequency change events from the original trace and generate start and end events. I'm not convinced it'd be simpler..

Reply:
yep, none of the simple slicing methods will result in the desired effect. I can get it (I think) by converting timestamps to actual DateTime objects, then reindexing using a new index generated by the intersection of the original index and the new range, but I think I prefer just iterating the frequency change events

Reply:
Yeah I was hoping to make our life easier with this but if that's not the case feel free to just ignore that :wink:

Reply:
Here's an example of recomputing the index, do you think it's cleaner?
```python
import bart

import pandas as pd

# 0.0->0.5 @ 450 -> 450 * 0.5 -> 225
# 0.5->1.5 @ 600 -> 600 * 1.0 -> 600
# 1.5->2.0 @ 900 -> 900 * 0.5 -> 450
# 2.0->2.5 @ 450 -> 450 * 0.5 -> 225
# 2.5->3.0 @ 600 -> 600 * 0.5 -> 300
# total 1800 / 3.0 -> 600 avg
# average between 0.25 and 2.25
# (112.5 + 600 + 450 + 112.5) / 2.0 -> 637.5
t=[0.0,0.5,1.5,2.0,2.5,3.0]
f=[450,600,900,450,600,450]
#series = pd.Series(data=f,index=pd.to_datetime(t,unit='s'))
series = pd.Series(data=f,index=t)
print "Complete Series:"
print series

# calculate the average frequency over a specific time range
start_time = 0.25
end_time   = 2.25
new_index = [ start_time ]+list(series[start_time:end_time].index)+[ end_time ]
filtered_series = series.reindex(index=new_index, method='pad')
print "\nSeries between {} and {}".format(start_time, end_time)
print filtered_series

print "Average Frequency for complete series: {}".format( bart.common.Utils.area_under_curve(series, method='rect') / (t[-1] - t[0]) )
print "Average Frequency for range: {}".format(bart.common.Utils.area_under_curve(filtered_series, method='rect') / (end_time - start_time))
```

```
Complete Series:
0.0    450
0.5    600
1.5    900
2.0    450
2.5    600
3.0    450
dtype: int64

Series between 0.25 and 2.25
0.25    450
0.50    600
1.50    900
2.00    450
2.25    450
dtype: int64
Average Frequency for complete series: 600.0
Average Frequency for range: 637.5
```

Reply:
The updated version LGTM :wink: